### PR TITLE
Expand and forward constructor arguments of H5Store to h5py.File.

### PR DESCRIPTION
**Pull request by csadorf**
## Description

 * Provide users with more options to control how the underlying h5py.File
   is constructed.
 * Implemented unit tests that test opening in read- and write-only mode.

## Motivation and Context
It was not possible to control the open mode prior to this patch.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

This pull request is based on
- [ ] *master*, because it is a bug fix or an update to the documentation.
- [x] *develop*, because it introduces a new feature.

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

### CODE REVIEW

File: signac/core/h5store.py [hunk 1]
```diff
@@ -261,16 +261,29 @@ class H5Store(MutableMapping):
             assert 'foo' in h5s
             assert h5s.foo == 'bar'
             assert h5s['foo'] == 'bar'
+
+    :param filename:
+        The filename of the underlying HDF5-file. This file will be create if it
```
Comment (bdice) [inline signac/core/h5store.py]: create ➡️ created

File: signac/core/h5store.py [hunk 2]
```diff
@@ -261,16 +261,29 @@ class H5Store(MutableMapping):
...
+        does not exist yet with the standard open mode ('a').
```
Comment (bdice) [inline signac/core/h5store.py]: Is this part about the standard open mode redundant? Or am I missing something?

File: signac/core/h5store.py [hunk 3]
```diff
@@ -298,7 +311,7 @@ def open(self):
         import h5py
         self._thread_lock.acquire()
         try:
-            self._file = h5py.File(self._filename)
+            self._file = h5py.File(self._filename, mode=self.mode, **self._kwargs)
```
Comment (bdice) [inline signac/core/h5store.py]: Should this be `self._mode`? edit: nevermind, I see the property, but why do we expose that property in the first place?

File: signac/core/h5store.py (hunk 2 above)
Comment (csadorf) [reply signac/core/h5store.py]: I wanted to make sure that users understand they don't *have* to create the file beforehand. The comment that 'a' is default may be redundant, but is that a problem?

File: signac/core/h5store.py (hunk 3 above)
Comment (csadorf) [reply signac/core/h5store.py]: It should be `_mode`. The mode is a property of the `H5Store`, why not expose it?

Comment (bdice) [reply signac/core/h5store.py]: I see it is exposed for file objects. We can expose it. https://docs.python.org/3/library/io.html#io.FileIO.mode

File: signac/core/h5store.py (hunk 2 above)
Comment (bdice) [reply signac/core/h5store.py]: To me, this implies that if you specify `mode='r'` and the file does not yet exist, it overrides your provided `'r'` and uses `'a'` instead. I'm not sure if that is true.

Comment (csadorf) [reply signac/core/h5store.py]: Got it, will revise.

**Comment by csadorf** [comment]
@bdice Is this better?

**Pull request merged**